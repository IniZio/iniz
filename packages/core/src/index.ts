export { atom, isAtom } from "./atom";
export { effect, Effect } from "./effect";
export { batch } from "./batch";
