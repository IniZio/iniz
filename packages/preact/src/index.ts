export * from "@iniz/core";
export { useAtom } from "./useAtom";
export { useComputed } from "./useComputed";
export { useSideEffect } from "./useSideEffect";
export { useStore } from "./useStore";
import "./createElement";
