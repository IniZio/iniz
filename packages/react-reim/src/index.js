import React, {PureComponent} from 'react'

import State from './components/state'

export {default as State} from './components/state'

// Syncs props to store
export function pipeTo(store, mutation = (_, p) => p) {
  class Piper extends PureComponent {
    componentDidUpdate() {
      store.setState(mutation, this.props)
    }

    render = () => null
  }
  return Piper
}

export const context = () =>
  ({
    name: 'context',
    call(store) {
      Object.defineProperties(store, {
        __isContext: {
          value: true
        },
        Consumer: {
          value: props => <State store={store} {...props}/>
        },
        get: {
          value: fn => <State store={store}>{fn}</State>
        }
      })
    }
  })

export const createContext = context()

export function connect(store, getter = s => s, setter = () => ({})) {
  const Context = store.__isContext ? store : store.plugin(context())
  return Wrapped => p => (
    <Context.Consumer getter={getter} setter={setter}>
      {
        selected => <Wrapped {...selected} {...p}/>
      }
    </Context.Consumer>
  )
}

export function useReim(store, getter = s => s) {
  const {useState, useEffect} = require('react')

  if (!useState) {
    throw new Error('React@16.7+ is required to use Hooks')
  }

  const [state, setState] = useState(store.snapshot(getter))

  useEffect(() => () => store.unsubscribe(setState))

  store.subscribe(setState, {getter})

  return [state, store.set]
}
