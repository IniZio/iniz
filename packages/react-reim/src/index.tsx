import React, {PureComponent} from 'react'
import reim, {Store, Mutation, Getter} from 'reim'

import State, {StateProps} from './components/state'

export {default as State} from './components/state'

class ReactStore extends Store {
  get __isReactReim() {return true}
  Consumer: (props: StateProps) => React.ReactElement<StateProps>;
  get: (fn: Getter) => React.ReactElement<StateProps>;
}

// Syncs props to store
export function pipeTo(store: Store, mutation: Mutation = (_: any, p: any) => p): React.ComponentClass {
  class Piper extends PureComponent {
    componentDidUpdate() {
      store.set(mutation, this.props)
    }

    render = () => null
  }
  return Piper
}

export const react = () =>
  ({
    name: 'react',
    call(store: Store) {
      Object.defineProperties(store, {
        __isReactReim: {
          value: true
        },
        Consumer: {
          value: (props: any) => <State store={store} {...props}/>
        },
        get: {
          value: (fn: Getter) => <State store={store} getter={fn}/>
        }
      })
    }
  })

export function connect(store: ReactStore, getter: Getter = s => s, setter: Mutation = () => ({})) {
  const Context = store.__isReactReim ? store : store.plugin(react())
  return (Wrapped: React.ComponentClass) => (p: any) => (
    <Context.Consumer getter={getter} setter={setter}>
      {
        selected => <Wrapped {...selected} {...p}/>
      }
    </Context.Consumer>
  )
}

export function useReim(initial: any, getter: Getter = s => s, dependencies: string[] = []) {
  // @ts-ignore
  if (!React.useState) {
    throw new Error('React@16.7-alpha.2 is required to use Hooks')
  }

  const store = initial.__isReim ? initial : reim(initial)

  // @ts-ignore
  const {useState, useEffect, useRef} = React

  const mountRef = useRef()
  const [state, set] = useState(store.snapshot(getter))

  useEffect(() => {
    if (mountRef.current) {
      set(() => store.snapshot(getter))
    } else {
      mountRef.current = true
    }
  }, dependencies)

  useEffect(() => () => store.unsubscribe(set))

  store.subscribe(set, {getter})

  return [state, store.set]
}

export default react
