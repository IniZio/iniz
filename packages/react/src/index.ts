export * from "@reim/core";

export { useAtom } from "./useAtom";
export { useSideEffect } from "./useSideEffect";
export { useCompute } from "./useCompute";