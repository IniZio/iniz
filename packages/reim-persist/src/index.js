import merge from 'lodash/merge'

function canWriteStorage(storage) {
  try {
    storage.setItem('@@', 1);
    storage.removeItem('@@');
    return true;
  } catch (e) {}

  return false;
}

function getState(key, storage, value) {
  try {
    return (value = storage.getItem(key)) && typeof value !== 'undefined'
      ? JSON.parse(value)
      : undefined;
  } catch (err) {}

  return undefined;
}

function setState(key, state, storage) {
  return storage.setItem(key, JSON.stringify(state));
}

export default function persist(options = {}) {
  const storage = options.storage || (window && window.localStorage)
  const key = `reim/${options.key}`

  if (!canWriteStorage(storage)) {
    throw new Error('Invalid storage given')
  }

  return {
    name: 'persist',
    apply(store) {
      const saved = getState(key, storage)
      if (typeof saved === 'object' && saved !== null) {
        store.setState(merge({}, store.state, saved))
      }

      store.subscribe(state => {
        setState(key, state, storage)
      }, {...options.subscriber, immediate: true})

      return store
    }
  }
}
