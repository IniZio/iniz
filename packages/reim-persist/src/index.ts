import {merge} from 'lodash'

function canWriteStorage(storage) {
  try {
    storage.setItem('@@', 1)
    storage.removeItem('@@')
    return true
  } catch (error) {}

  return false
}

function snapshot(key, storage, value?: any) {
  try {
    value = storage.getItem(key)
    return typeof value === 'undefined' ? undefined : JSON.parse(value)
  } catch (error) {}

  return undefined
}

function set(key, state, storage) {
  return storage.setItem(key, JSON.stringify(state))
}

export default function persist(options: any = {}) {
  const storage =
    options.storage ||
    (typeof window === 'undefined' ? null : window && window.localStorage)

  if (!canWriteStorage(storage)) {
    throw new Error('Invalid storage given')
  }

  return (store): void => {
    // console.log('persist store? ', store)

    if (!store.name || store.name.length <= 0) {
      throw new Error(
        'You cannot persist an anonymous store, use `name` option in store'
      )
    }

    const key = `reim/${store.name}`
    const saved = snapshot(key, storage)
    if (typeof saved === 'object' && saved !== null) {
      store.reset(merge({}, store._state, saved))
    }

    store.subscribe(
      state => {
        set(key, state, storage)
      },
      {...options.subscriber, immediate: true}
    )
  }
}
