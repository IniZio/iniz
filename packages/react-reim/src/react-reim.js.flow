// @flow
import {Component} from 'react'

import {Mutation, Getter, Plugin, Store, State} from 'reim'

export type Setter = State | {(state: State):  void | State}

export function pipeTo(store: Store, mutation: Mutation): Component

export function context(): Plugin
export const createContext: Plugin

export const State: Component

export function connect(store: Store, getter: Getter): {(Wrapped: Component): Component}
