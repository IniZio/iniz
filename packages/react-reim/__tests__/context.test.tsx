/* eslint react/prop-types: 0 */
/* eslint-disable-next-line import/no-extraneous-dependencies  */
import * as renderer from 'react-test-renderer'
import * as React from 'react'
import reim from 'reim'
import {react, connect, State} from '../src'

test('react returns Consumer', () => {
  const store = reim({yer: 43}).plugin(react())

  expect(store.Consumer).toBeDefined()
})

test('Storeless State should reset on initial prop change', () => {
  let changeInitial

  function TestComponent() {
    const [initial, setInitial] = React.useState({level: 10})

    changeInitial = () => setInitial({level: 10000})

    return (
      <State initial={initial}>
        {({level}) => (
          <h1>{level}</h1>
        )}
      </State>
    )
  }

  const component = renderer.create(<TestComponent/>)
  expect(component.toJSON()).toMatchSnapshot()

  changeInitial()
  expect(component.toJSON()).toMatchSnapshot()
})

test('Consumer should have change in store state reflected', () => {
  const store = reim({yer: 43}).plugin(react())

  const component = renderer.create(
    <store.Consumer>
      {
        state => (
          <div>
            <div id="value">{state.yer}</div>
          </div>
        )
      }
    </store.Consumer>
  )
  store.set(state => {
    state.yer += 88
  })

  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Existing store Consumer should NOT be able to set initial value', () => {
  const store = reim({yer: 9}).plugin(react())

  const component = renderer.create(
    <store.Consumer initial={{yer: 10}}>
      {
        state => (
          <div>
            <div id="value">{state.yer}</div>
          </div>
        )
      }
    </store.Consumer>
  )

  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('get should have change in store state reflected', () => {
  const {get, set} = reim({yer: 43}).plugin(react())

  const component = renderer.create(
    get(state => (
      <div>
        <div id="value">{state.yer}</div>
      </div>
    ))
  )

  const component1 = renderer.create(get('yer'))

  set(state => {
    state.yer += 88
  })

  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()

  expect(component1.toJSON()).toMatchSnapshot()
})

test('Unmount Cunsumer should unsubscribe', () => {
  const store = reim({yer: 43}).plugin(react())

  const getter = jest.fn()

  const component = renderer.create(
    <store.Consumer getter={getter}>
      {
        state => (
          <div>
            <div id="value">{state.yer}</div>
          </div>
        )
      }
    </store.Consumer>
  )
  store.set(state => {
    state.yer += 88
  })
  expect(getter).toBeCalledTimes(2)
  component.unmount()
  store.set(state => {
    state.yer *= 88
  })
  expect(getter).toBeCalledTimes(2)
})

test('State component should work', () => {
  const component = renderer.create(
    <State initial={{mm: 88}}>
      {
        ({mm}) => (
          <h1>{mm}</h1>
        )
      }
    </State>
  )

  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('State onChange should trigger on state change', () => {
  const onChange = jest.fn()
  let increment

  class Container extends React.Component {
    render() {
      return (
        <State initial={{value: 88}} onChange={onChange}>
          {
            ({value}, {set}) => {
              increment = () => set({value: value + 1})
              return <h1>{value}</h1>
            }
          }
        </State>
      )
    }
  }

  const component = renderer.create(<Container/>)

  expect(onChange).toBeCalledTimes(0)

  increment()

  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()

  expect(onChange).toBeCalledTimes(1)
})

test('Properties not included in getter should not trigger update', () => {
  const store = reim({hel: 43, gee: 10}).plugin(react())

  const updated = jest.fn()

  class Listen extends React.Component<{hel: any}> {
    componentDidUpdate = updated

    render() {
      return <div>{this.props.hel}</div>
    }
  }

  renderer.create(
    <store.Consumer getter={state => ({hel: state.hel})}>
      {
        state => (
          <Listen {...state}/>
        )
      }
    </store.Consumer>
  )
  store.set(state => {
    state.hel -= 22
  })
  expect(updated).toBeCalledTimes(1)
  store.set(state => {
    state.gee += 88
  })
  expect(updated).toBeCalledTimes(1)
})

test('use convenience method connect', () => {
  const store = reim({bom: 19})

  const Container = connect(store, state => ({bom: state.bom}))(
    state => <div>{JSON.stringify(state)}</div>
  )

  const component = renderer.create(
    <Container/>
  )
  store.set(state => {
    state.bom += 490
  })
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('change getter', () => {
  const store = reim({dui: 12, geo: 'tie'}).plugin(react())

  class Getter extends React.Component {
    state = {
      getter: state => ({dui: state.dui})
    }

    render() {
      return (
        <store.Consumer getter={this.state.getter}>
          {
            state => (
              <div>
                <div id="selected">{JSON.stringify(state)}</div>
              </div>
            )
          }
        </store.Consumer>
      )
    }
  }

  const component = renderer.create(<Getter/>)
  expect(component.toJSON()).toMatchSnapshot()

  // Change in getter should change passed in state
  const {instance} = component.root
  instance.setState({getter: s => s})
  expect(component.toJSON()).toMatchSnapshot()
})
