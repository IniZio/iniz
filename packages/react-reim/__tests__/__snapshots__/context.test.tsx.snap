// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Consumer should have change in store state reflected 1`] = `
<div>
  <div
    id="value"
  >
    131
  </div>
</div>
`;

exports[`Existing store Consumer should NOT be able to set initial value 1`] = `
<div>
  <div
    id="value"
  >
    9
  </div>
</div>
`;

exports[`State component should work 1`] = `
<h1>
  88
</h1>
`;

exports[`State onChange should trigger on state change 1`] = `
<h1>
  89
</h1>
`;

exports[`Storeless State should reset on initial prop change 1`] = `
<h1
  onClick={[Function]}
>
  10
</h1>
`;

exports[`Storeless State should reset on initial prop change 2`] = `
<h1
  onClick={[Function]}
>
  10
</h1>
`;

exports[`change filter 1`] = `
<div>
  <div
    id="selected"
  >
    {"dui":12}
  </div>
</div>
`;

exports[`change filter 2`] = `
<div>
  <div
    id="selected"
  >
    {"dui":12}
  </div>
</div>
`;

exports[`use convenience method connect 1`] = `
<div>
  {"bom":509}
</div>
`;
